### PR TITLE
Added logstash storage driver

### PR DESCRIPTION
**Pull request by joeyasperger**
Added a Logstash storage driver. This is useful because the current Elasticsearch storage driver does not offer very much flexibility. A Logstash storage driver allows developers to process data before storing it in Elasticsearch.

For my use case in particular, this storage driver allows for me to use Logstash to take derivatives of various stats like CPU usage before storing the data in Elasticsearch , which makes it much easier to make useful visualizations. I also use Logstash to break up the records into smaller Elasticsearch events, which again is useful when making visualizations. Lastly, I can also create timestamped indices in Logstash, unlike the single "cadvisor" index created by the Elasticsearch storage driver, which makes it easier to manage my indices.

### CODE REVIEW

**Comment by joeyasperger** [comment]
I signed it!

**Comment by magnusbaeck** [comment]
Why call this storage driver Logstash? It just serializes the data to JSON and pushes it over UDP. That could easily be used with other tools than Logstash.

**Comment by joeyasperger** [comment]
@magnusbaeck Yeah I though about giving it a more general name and you're right I probably should've done that. I'll update it later today.

**Comment by GabKlein** [comment]
Any update on this PR ?

**Comment by j00p34** [comment]
Is this going to be merged. Really need this functionality. @timstclair ?

**Comment by timstclair** [comment]
I was thinking this was blocked by https://github.com/google/cadvisor/issues/1458, but after skimming I like that it is a generic push-model, so I think it fits into the current design more cleanly. I'll try to get to a review in the next week or so.

File: storage/json/json.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+	containerName := ref.Name
+	if len(ref.Aliases) > 0 {
+		containerName = ref.Aliases[0]
+	}
+	// convert stats into json object
+	output, err := json.Marshal(DetailSpec{self.infoField, self.machineName, containerName, stats})
+	if err != nil {
+		return err
+	}
+	// write json object to udp connection
```
Comment (dashpole) [inline storage/json/json.go]: nit: remove "udp", or add "udp/tcp" since protocol is configurable

File: storage/json/json.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+	_, err = self.connection.Write(output)
+	return err
+}
+
+func (driver *jsonStorage) Close() error {
+	return nil
+}
+
+func newStorage(machineName string, storageHost string, infoField string, protocol string) (*jsonStorage, error) {
+	// create udp connection to host
```
Comment (dashpole) [inline storage/json/json.go]: nit: same as above

File: storage/json/json.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+	if err != nil {
+		return err
+	}
+	// write json object to udp connection
+	_, err = self.connection.Write(output)
+	return err
+}
+
+func (driver *jsonStorage) Close() error {
+	return nil
```
Comment (dashpole) [inline storage/json/json.go]: You should close connections:  "return driver.connection.Close()".  Correct me if I am missing something.

**Comment by timstclair** [comment]
Superseded by https://github.com/google/cadvisor/pull/1575

**Pull request closed without merging**